diff --git a/mysql-v5.7.32/mysqldump.c b/mysql-v5.7.32/mysqldump.c
index 9753f56..f0aef22 100644
--- a/mysql-v5.7.32/mysqldump.c
+++ b/mysql-v5.7.32/mysqldump.c
@@ -121,6 +121,7 @@ static my_bool  verbose= 0, opt_no_create_info= 0, opt_no_data= 0,
                 opt_events= 0, opt_comments_used= 0,
                 opt_alltspcs=0, opt_notspcs= 0, opt_drop_trigger= 0,
                 opt_secure_auth= TRUE;
+static my_bool opt_exec_only_in_mysql = 0;                
 static my_bool insert_pat_inited= 0, debug_info_flag= 0, debug_check_flag= 0;
 static ulong opt_max_allowed_packet, opt_net_buffer_length;
 static MYSQL mysql_connection,*mysql=0;
@@ -570,6 +571,9 @@ static struct my_option my_long_options[] =
    "Enable/disable the clear text authentication plugin.",
    &opt_enable_cleartext_plugin, &opt_enable_cleartext_plugin,
    0, GET_BOOL, OPT_ARG, 0, 0, 0, 0, 0, 0},
+  {"exec-only-in-mysql", 0, "Enable sequoiadb_execute_only_in_mysql.",
+   &opt_exec_only_in_mysql, &opt_exec_only_in_mysql, 0,
+   GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},  
   {0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
 };
 
@@ -5981,6 +5985,8 @@ static my_bool get_view_structure(char *table, char* db)
   FILE       *sql_file= md_result_file;
   my_bool    freemem= FALSE;
   char const *text;
+  char       table_name_buff[NAME_LEN*2+3] = {0};
+  char       db_name_buff[NAME_LEN*2+3] = {0};
   DBUG_ENTER("get_view_structure");
 
   if (opt_no_create_info) /* Don't write table creation info */
@@ -6033,11 +6039,15 @@ static my_bool get_view_structure(char *table, char* db)
   verbose_msg("-- Dropping the temporary view structure created\n");
   fprintf(sql_file, "/*!50001 DROP VIEW IF EXISTS %s*/;\n", opt_quoted_table);
 
+  mysql_real_escape_string(mysql, table_name_buff, table, (ulong)strlen(table));
+  mysql_real_escape_string(mysql, db_name_buff, db, (ulong)strlen(db));
+
   my_snprintf(query, sizeof(query),
               "SELECT CHECK_OPTION, DEFINER, SECURITY_TYPE, "
               "       CHARACTER_SET_CLIENT, COLLATION_CONNECTION "
               "FROM information_schema.views "
-              "WHERE table_name=\"%s\" AND table_schema=\"%s\"", table, db);
+              "WHERE table_name=\"%s\" AND table_schema=\"%s\"",
+              table_name_buff, db_name_buff);
 
   if (mysql_query(mysql, query))
   {
@@ -6298,6 +6308,12 @@ int main(int argc, char **argv)
   if (opt_single_transaction && do_unlock_tables(mysql)) /* unlock but no commit! */
     goto err;
 
+  if (opt_exec_only_in_mysql) {
+    const char *query = "SET SESSION sequoiadb_execute_only_in_mysql = 1";
+    if (mysql_query_with_error_report(mysql, 0, query)) 
+      goto err;
+  }
+ 
   if (opt_alltspcs)
     dump_all_tablespaces();
 
