diff --git a/mysql-v5.7.25/mysqldump.c b/mysql-v5.7.25/mysqldump.c
index 4c610e6..5b4a8be 100644
--- a/mysql-v5.7.25/mysqldump.c
+++ b/mysql-v5.7.25/mysqldump.c
@@ -114,6 +114,7 @@ static my_bool  verbose= 0, opt_no_create_info= 0, opt_no_data= 0,
                 opt_events= 0, opt_comments_used= 0,
                 opt_alltspcs=0, opt_notspcs= 0, opt_drop_trigger= 0,
                 opt_secure_auth= TRUE;
+static my_bool opt_exec_only_in_mysql = 0;
 static my_bool insert_pat_inited= 0, debug_info_flag= 0, debug_check_flag= 0;
 static ulong opt_max_allowed_packet, opt_net_buffer_length;
 static MYSQL mysql_connection,*mysql=0;
@@ -563,6 +564,9 @@ static struct my_option my_long_options[] =
    "Enable/disable the clear text authentication plugin.",
    &opt_enable_cleartext_plugin, &opt_enable_cleartext_plugin,
    0, GET_BOOL, OPT_ARG, 0, 0, 0, 0, 0, 0},
+  {"exec-only-in-mysql", 0, "Enable sequoiadb_execute_only_in_mysql.",
+   &opt_exec_only_in_mysql, &opt_exec_only_in_mysql, 0,
+   GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
   {0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
 };
 
@@ -6286,6 +6290,12 @@ int main(int argc, char **argv)
   if (opt_single_transaction && do_unlock_tables(mysql)) /* unlock but no commit! */
     goto err;
 
+  if (opt_exec_only_in_mysql) {
+    const char *query = "SET SESSION sequoiadb_execute_only_in_mysql = 1";
+    if (mysql_query_with_error_report(mysql, 0, query))
+      goto err;
+  }
+
   if (opt_alltspcs)
     dump_all_tablespaces();
 
@@ -6383,3 +6393,4 @@ err:
 
   return(first_error);
 } /* main */
+
