--- sql_parse.cc	2020-06-29 10:14:35.918248603 +0800
+++ sql_parse_modified.cc	2020-06-29 10:10:04.266255923 +0800
@@ -4992,6 +4992,15 @@
       thd->killed= THD::NOT_KILLED;
     }
   }
+#ifndef EMBEDDED_LIBRARY
+  else {
+    mysql_audit_notify(thd,
+                      first_level ? MYSQL_AUDIT_QUERY_STATUS_END :
+                                    MYSQL_AUDIT_QUERY_NESTED_STATUS_END,
+                      first_level ? "MYSQL_AUDIT_QUERY_STATUS_END" :
+                                    "MYSQL_AUDIT_QUERY_NESTED_STATUS_END");
+  }
+#endif /* !EMBEDDED_LIBRARY */
 
   lex->unit->cleanup(true);
   /* Free tables */
